### ISSUE TITLE
nullable & optional/required marks

### ISSUE THREAD

**Issue by raderio** [Kotlin/kotlinx.serialization#407]
Kotlin solves NullPointerException with nullable/non-nullable classes.

But sometimes we need to show that a field is required or not. Yes sometimes we can express this by using null, but not always.
An use case will be partial update, when we want to update just some columns in database.
So, if a column is nullable in database, we can not send null to that field, because null is valid value for it.

A solution will be to use a class `Maybe` or `Option`, something like java's `Optional`.
The issue with `Optional` is that it accepts null even if we have `Optional<String>`, it should accept null only if we have `Optional<String?>`.

1. Option<String> - optional & non-nullable
2. Option<String?> - optional & nullable
3. String - required & non-nullable
4. String? - required & nullable

This mean, or we send a data for the field, or just do not send the field in request, null is not allowed.
`{"foo": "bar", "baz": true}` is allowed
`{"baz": true}` is allowed, because is optional
`{"foo": null, "baz": true}` is NOT allowed, because is non-nullable

Basically `null` has the meaning of "this property existed before, and I want to nullify its value". If there's no value, or the value needs not change (for requests, for example), it just shouldn't be sent.

An example of partial update you can find here https://medium.com/workflowgen/graphql-mutations-partial-updates-implementation-bff586bda989
This is also supported by OpenAPI https://swagger.io/docs/specification/data-models/data-types/ you can find that a filed can be marked as optional & non-nullable

**Comment by sandwwraith** [comment]
As far as I understand your feature request, it is already supported – if you mark `foo` as `@Optional` and it has a non-nullable type, absence of value will be allowed and nulls will not.

However, it still has to have a default value. The functionality of 'partial patches' or 'partial views' is not planned for 1.0 release, but we may consider it later.

**Comment by raderio** [comment]
```
class Req(@Optional val foo: String)
```
Is it possible? 
If so, what value will have `foo` if the request will be empty `{}` ?

**Comment by sandwwraith** [comment]
When you use `@Optional`, compiler checks if `foo` has a default value. So code above will not compile, and this will do:

`class Req(@Optional val foo: String = "DEFAULT")`

**Comment by raderio** [comment]
Ok, but in my case I do not have a default value, and I can not put a random string as a default, because in that case I will not be able to make difference, is it my default or this value was in the request.

So, my solution is `class Req(val foo: Option<String>)`, or value is provided or is missing, null is not allowed.

**Comment by raderio** [comment]
Adding `Option` class and in case of missing property on deserialization set `None` value will solve this issue.
What do you think? I can help with pull request in case it is approved.

**Comment by sandwwraith** [comment]
As far as I know, there is no `Option` class in Kotlin multiplatform stdlib, so you can end up creating you own class, because there is no need in modification of library.

**Comment by raderio** [comment]
Yes, there is not `Option` class in Kotlin stdlib, but maybe it is possible to add it into this library?
